photologue/utils/zipfile.py

```diff
@@ -0,0 +1,114 @@
+import os
+from io import BytesIO
+from django.conf import settings
+from django.template.defaultfilters import slugify
+from django.core.files.base import ContentFile
+from PIL import Image
+from django.contrib.sites.models import Site
+import zipfile
+from photologue.models import Photo, Gallery
+import logging
+from django.contrib import messages
+from django.utils.translation import ugettext_lazy as _
+from django.utils.encoding import force_text
+
+logger = logging.getLogger('photologue.zip')
+
+# Use celery to speed up page loading after uploading photos?
+USE_CELERY = getattr(settings, 'PHOTOLOGUE_USE_CELERY', False)
+
+
+def parse_zip(zip_file, gallery, title, description, caption, is_public, request=None):
+    zip = zipfile.ZipFile(zip_file)
+    count = 1
+    current_site = Site.objects.get(id=settings.SITE_ID)
+    if gallery:
+        logger.debug('Using pre-existing gallery.')
+    else:
+        logger.debug(
+            force_text('Creating new gallery "{0}".').format(title))
+        gallery = Gallery.objects.create(title=title,
+                                         slug=slugify(title),
+                                         description=description,
+                                         is_public=is_public)
+        gallery.sites.add(current_site)
+
+    for filename in sorted(zip.namelist()):
+
+        logger.debug('Reading file "{}".'.format(filename))
+
+        if filename.startswith('__') or filename.startswith('.'):
+            logger.debug('Ignoring file "{}".'.format(filename))
+            continue
+
+        if os.path.dirname(filename):
+            logger.warning('Ignoring file "{}" as it is in a subfolder; all images should be in the top '
+                           'folder of the zip.'.format(filename))
+            if request:
```

Comment:
At present we call the UploadZipForm from the admin, so there's always a `request` object passed as an argument. So if there's a problem, the user receives feedback on screen. 
How do you propose to warn the user if there is an issue when we are using Celery? Is there a mechanism in Celery to provide exit values, that we could potentially read and display?

Reply:
I thought about this as well and I thought about 3 solutions, which I think all aren't the greatest

- Using channels to send error messages to the user. Two reasons why I don't really like this idea: 
    1. It is unnecessarily complex and introduces a new dependency on django channels
    2. Messages are only received if the user is on the right page at the right time if a developer only implements the default pages. Although he could subscribe to the channel on every page of his application if he so chooses. In that case, this would bring the best end-user experience I think.

- Saving errors to the database and then we've got 2 ways to show these errors to the user
    1. Writing some kind of custom middleware to read these from the database and show messages if there are any (not a great solution, this means two extra db-calls for every page-load of every user -> one for retrieving and one for deleting)
    2. Only showing these error-messages on some pages used in the admin-dashboard. And if the developer wants to show these messages on other pages as well, he'll have to add the messages on these pages himself. (we could define some function show_upload_errors_as_messages(request))

The last solution seems the best of the 3 solutions according to me. But it'd be great if someone had an even better idea.

Reply:
Thanks for the feedback. I agree that none of the solutions are great. It's somewhat complicated because we don't know what is installed on the Django project in which Photologue will be used. If the project already uses channels - great, use channels. Otherwise, it's not a good idea.
I like your suggestion of saving async feedback messages in a database table and showing them in the admin. This also gives the developer flexibility: if they want to use middleware to display feedback messages, it's easy to implement.
A suggestion: if the developer plans to use channels: would it be useful to put a `hook` in the Photologue messages code, so that a developer can add a custom output (e.g. channels) if they wish for their project?